rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_publisher.hpp

```diff
@@ -96,8 +96,11 @@ class LifecyclePublisher : public LifecyclePublisherInterface,
    * was enabled or disabled and forwards the message
    * to the actual rclcpp Publisher base class
    */
+  [[deprecated(
+    "publishing an unique_ptr is prefered when using intra process communication."
+    " If using a shared_ptr, use publish(*msg).")]]
   virtual void
-  publish(const std::shared_ptr<const MessageT> & msg)
+  publish(std::shared_ptr<const MessageT> msg)
```

Comment:
This is quite strange:
If I overload the publish method here, don't add a `deprecated` attribute and call a non-deprecated publish method in the base-class -> I never see a warning.
If I overload the publish method here, add a `deprecated` attribute and call a non-deprecated publish method in the base-class -> I see a warning in template specialization.
If I overload the publish method here, don't add a `deprecated` attribute and call the deprecated publish method in the base-class -> I see a warning when the package is being built, but not when the dependencies are built.
(Ubuntu 18.04, gcc)

As a workaround, I'm using `const std::shared_ptr<const MessageT> & msg` in the base class and `std::shared_ptr<const MessageT> msg` here.

If I deprecate the publish methods taking a ptr, I will have a similar problem. @wjwwood suggestions?

Maybe I could directly delete them from the lifecycle publisher, which has less packages depending on it.

Reply:
And the trick doesn't work, because it generates a different [warning in `macOS`](https://ci.ros2.org/job/ci_osx/5629/warnings11Result/package.-1256088273/).
I don't see a way of deprecating a method in the base class and the overloading method in the derived class without generating a warning, I will directly delete the `deprecated` methods in `lifecycle_publisher`, it shouldn't generate much problems.